src/core/ext/filters/client_channel/resolver/fake/fake_resolver.cc

```diff
@@ -95,43 +99,48 @@ FakeResolver::~FakeResolver() {
   grpc_channel_args_destroy(channel_args_);
 }
 
-void FakeResolver::NextLocked(grpc_channel_args** target_result,
-                              grpc_closure* on_complete) {
-  GPR_ASSERT(next_completion_ == nullptr);
-  next_completion_ = on_complete;
-  target_result_ = target_result;
-  MaybeFinishNextLocked();
+void FakeResolver::StartLocked() {
+  active_ = true;
+  MaybeSendResultLocked();
 }
 
 void FakeResolver::RequestReresolutionLocked() {
   if (reresolution_results_ != nullptr || return_failure_) {
     grpc_channel_args_destroy(next_results_);
     next_results_ = grpc_channel_args_copy(reresolution_results_);
-    MaybeFinishNextLocked();
+    // Return the result in a different closure, so that we don't call
```

Comment:
If this is something should be generally required of any resolver, does it make sense to push this responsibility of post-poning the LB policy callback into the resolver result handler? I'd guess the fact that this explicit bounce isn't needed in other resolvers is probably because those resolvers just accidentally happen to complete their resolutions  using a closure sched somewhere along the way.

Reply:
Any resolver that is actually doing a resolution will need to do I/O, which means that its result will always be returned asynchronously, so it will never call any method on the result handler from within the `RequestReresolutionLocked()` method.  The fake resolver is special in this regard, so we don't need a more general solution.

That having been said, I have added a comment indicating that resolver implementations must not call the result handler from within the `RequestReresolutionLocked()` method.

Reply:
This should work either way. But note, for example, that there are several situations in which the c-ares resolver doesn't actually perform any async I/O for a resolution - for example if the target is an ip literal, or if the target can be found in the hosts file. In either of those cases, that resolver does use a closure sched but only because that's how the the current "on done" from the wrapper is written: https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L154. I'd imagine other resolvers could have similar cases.

What do you think?

Reply:
I think those are special cases that can be handled by the resolver implementation.  If a resolver is special-casing some request such that it doesn't need to do I/O, it can be responsible for returning the result in a closure.

There are two reasons that I'd prefer not to put this in the result handler.  First, it would make the API a bit ugly by making it such that the method defined by the result handler implementation is different from the one called by the resolver.  In other words, it would need to look something like this:

```
class ResultHandler {
 public:
  ResultHandler() {
    GRPC_CLOSURE_INIT(&closure_, ReturnResultLocked, this,
            grpc_combiner_scheduler(combiner()));
  }

  void ReturnResult(const grpc_channel_arg* result) {
    current_result_ = result;
    GRPC_CLOSURE_SCHED(&closure_, GRPC_ERROR_NONE);
  }

 protected:
  // Subclasses implement this.
  virtual void ReturnResultImpl(const grpc_channel_args* result) GRPC_ABSTRACT;

 private:
  static void ReturnResultLocked(void* arg, grpc_error* error) {
    ResultHandler* self = static_cast<ResultHandler*>(arg);
    self->ReturnResultImpl(self->current_result_);
    self->current_result_ = nullptr;
  }

  const grpc_channel_args* current_result_ = nullptr;
  grpc_closure closure_;
};
```

The second problem is that it would require an additional memory allocation for each call to `ReturnResult()`; we could not simply use a closure allocated as a member of the result handler as shown above, because there is no guarantee that the previous result closure would be executed before the next one is requested.

Now, this is on the control plane, not the data plane, so I am willing to go that route if we need to at some point in the future.  But until and unless we have a demonstrated need to do that, I would prefer to stick with making it the resolver's responsibility to deal with these special cases.